City-of-Helsinki/events-helsinki-monorepo
# HH-254: Add error handler for Apollo link with custom error page and use ErrorBoundary for render errors.

## Description
There is three ways to handle errors: 
- show 404 / 500  page
- show _error page
- show error boundary component (which is basically same as the 500 page)

404 / 500 pages are shown for server side errors. _error-page is shown when static props revalidation throws error. ErrorBoundary pops up when render method throws error. However there were some problems along the way:

1. First off, the ErrorBoundary was in a wrong place. Because ErrorBoundary component still needs to show header and footer sections, it needs to be inside the ApolloProvider and CMS Provider, not the outmost part of the _app.
2. `500.tsx `didnt work, it always failes to getStaticProps. The reason is a bug in nextjs when middleware.ts is implemented. There was an actually an open issue for that: https://github.com/vercel/next.js/issues/38762
There is no fix for tha, so only thing to do was to instantly redirect to some different page, so it will always redirect now to the `error.tsx `page which mimics the original 500 page.
3.  `_error.tsx` translations were broken on _error page: https://github.com/i18next/next-i18next/issues/1020
There is no fix either but we can use the same redirect hack here and go to `error.tsx` page.



```
const {data, loading, error } = useSomeQuery(...):
useErrorBoundary(error);
```
Then the `useErrorBoundary` hook will trigger the error boundary which fallbacks to error boundary component (500 page).
 Errorboundary is implemented with `react-error-boundary` library. There is good blog post about it how it works, and why is it good choice for error handling: https://kentcdodds.com/blog/use-react-error-boundary-to-handle-errors-in-react
 
 It's quite hard to test whether the boundary triggers, because boundary works only in production mode. In dev mode the nextjs error console overrides the error handler.  I found two ways to manually test it. 
 1) Trigger it when user clicks some button (example in the blog post above) or 
 2) in useEffect like this:
 
Wohoo, everything works now correctly!

how to test:

1. start script to run the apollo server locally yarn docker:graphql-router:hobbies:serve from root of the monorepo
2. run yarn build-fast under app/sports-helsinki
3. run yarn start
4. go to: http://localhost:3000 and browse around
5. stop the apollo server
6. browse some more
-> error.tsx page should pop up

PS: The apollo link handler is only done for sports-helsinki. Errorboundary works for other services too. This apollo link error handler POC code has to be refactored so that it can be used in all three services. This can be done in next pr.

PPS: Some decen refactoring done because I hate copy-pasting three times. next has now common config in the root file called next.base.config.js. Any config can be overridden from project, but currently it seems that every project uses same config. Also All moving parts from `_App.tsx` root app files are moved to `BaseApp.tsx` component under components folder.


## Issues

### Closes

**[DEV-XXX](https://helsinkisolutionoffice.atlassian.net/browse/DEV-XXX):**

### Related

## Testing

### Automated tests

### Manual testing

## Screenshots

## Additional notes
